src/java/com/puppetlabs/puppetserver/pool/JRubyPool.java

```diff
@@ -1,234 +1,371 @@
     }
 
-    /**
-    * This method is analogous to <tt>putFirst</tt> in the
-    * <tt>LinkedBlockingDeque</tt> class, but also causes a read lock to be
-    * unlocked from the <tt>ReentrantReadWriteLock</tt>.  This method should
-    * only be called from the same thread that was running when the item was
-    * originally borrowed.
-    *
-    * @param e the element to return to the pool
-    * @throws InterruptedException
-    */
     @Override
-    public void releaseItem(E e) throws InterruptedException {
+    public void releaseItem(E e) {
         releaseItem(e, true);
     }
 
-    /**
-    * This method is analogous to <tt>putFirst</tt> in the
-    * <tt>LinkedBlockingDeque</tt> class, but also causes a read lock to be
-    * unlocked from the <tt>ReentrantReadWriteLock</tt>.  This method should
-    * only be called from the same thread that was running when the item was
-    * originally borrowed.
-    *
-    * @param e the element to return to the pool
-    * @param returnToPool whether to return the element to the pool (true)
-    *                     or just throw the element away (false)
-    * @throws InterruptedException
-    */
     @Override
-    public void releaseItem(E e, boolean returnToPool)
-            throws InterruptedException {
+    public void releaseItem(E e, boolean returnToPool) {
+        final ReentrantLock lock = this.lock;
+        lock.lock();
         try {
             if (returnToPool) {
-                liveQueue.putFirst(e);
+                addFirst(e);
             }
         } finally {
-            lock.readLock().unlock();
+            lock.unlock();
         }
     }
 
     /**
-    * This method is analagous to <tt>putFirst</tt> in the
-    * <tt>LinkedBlockingDeque</tt> class. It should only ever be used to
-    * insert a `PoisonPill` or `RetryPoisonPill` to the pool, which is an
-    * operation that should be done without acquiring a read lock.
-    *
-    * @param e the element to add to the queue
-    * @throws InterruptedException
-    */
+     * Insert a poison pill into the pool.  It should only ever be used to
+     * insert a `PoisonPill` or `RetryPoisonPill` to the pool.
+     */
     @Override
-    public void insertPill(E e) throws InterruptedException {
-        liveQueue.putFirst(e);
+    public void insertPill(E e) {
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            addFirst(e);
+        } finally {
+            lock.unlock();
+        }
     }
 
-    /**
-    * This method is analagous to <tt>clear</tt> in the
-    * <tt>LinkedBlockingDeque</tt> class.
-    *
-    * @throws InterruptedException
-    */
     @Override
     public void clear() {
-        liveQueue.clear();
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            int queueSize = liveQueue.size();
+            for (int i=0; i<queueSize; i++) {
+                registeredElements.remove(liveQueue.removeFirst());
+            }
```

Comment:
@cprice404 also mentioned this here - https://github.com/puppetlabs/puppet-server/commit/36ccd4e6cb3ceb9ddf408aea5d894d81095af7b9#commitcomment-14854753.

Basically, I was thinking of the scenario where an instance may have been borrowed from the `liveQueue` at the time `clear` was called.  It would seem weird to me to have `clear` remove the associated element from `registeredElements` for the currently borrowed item and later possibly have that borrowed item inserted back into the `liveQueue` with no associated element available in `registeredElements` anymore.  There are probably a number of different approaches for dealing with this, but I thought it might be simplest to only remove the `registeredElements` that are likewise being removed from the `liveQueue`.

FWIW, I don't think the `clear` method is even used in any production code and the only test code reference I'm aware of is in the test I added for it in this PR.  I only kept it here because it exists in current code.

Does that make sense?